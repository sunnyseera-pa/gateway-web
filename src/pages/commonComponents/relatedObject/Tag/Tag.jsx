/** @jsx jsx */
import { jsx } from '@emotion/react';
import React from 'react';
import PropTypes from 'prop-types';
import { toTitleCase } from '../../../../utils/GeneralHelper.util';
import * as styles from '../Dataset/Dataset.styles';
import '../../CommonComponents.scss';

const Tag = props => {
	const { tagName, tagType, activeLink, onSearchPage, updateOnFilterBadgeHandler, parentKey, filter, url, showTagType, version } = props;
	const displayTagName = showTagType ? (
		<>
			<span>{`${toTitleCase(tagType)}: ${tagName}`}</span>
			<span>{version}</span>
		</>
	) : (
		<>
			<span>{tagName}</span>
			<span>{version}</span>
		</>
	);
	if (activeLink) {
		if (onSearchPage) {
			return (
				<span
					css={styles.pointer}
					onClick={event => updateOnFilterBadgeHandler(filter, { label: tagName, parentKey: parentKey })}
					data-testid={`badge-${tagName}-span`}>
					<div className={`badge-${tagType}`} data-testid={`badge-${tagName}`}>
						{displayTagName}
					</div>
				</span>
			);
		} else {
			return (
				<a href={`${url}${tagName}`} data-testid={`badge-${tagName}-link`}>
					<div className={`badge-${tagType}`} data-testid={`badge-${tagName}`}>
						{displayTagName}
					</div>
				</a>
			);
		}
	} else {
		return (
			<div className={`badge-${tagType}`} data-testid={`badge-${tagName}`}>
				{props.children}
				{displayTagName}
			</div>
		);
	}
};

Tag.propTypes = {
	tagName: PropTypes.string.isRequired,
	tagType: PropTypes.string.isRequired,
	activeLink: PropTypes.bool.isRequired,
	onSearchPage: PropTypes.bool.isRequired,
	parentKey: PropTypes.string.isRequired,
	filter: PropTypes.string.isRequired,
	url: PropTypes.string.isRequired,
	updateOnFilterBadgeHandler: PropTypes.func.isRequired,
	showTagType: PropTypes.bool.isRequired,
	version: PropTypes.string,
};
Tag.defaultProps = {
	version: '',
	filter: '',
};

export default Tag;
