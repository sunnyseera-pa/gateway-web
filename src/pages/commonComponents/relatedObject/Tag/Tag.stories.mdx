import { Meta, Story, Canvas } from '@storybook/addon-docs';
import Tag from './Tag';

<Meta
	title='Related Resources/Tag'
	component={Tag}
	args={{
		tagName: 'test',
		tagType: 'phenotype',
		onSearchPage: true,
		activeLink: true,
		url: '/search?search=&tab=Datasets&phenotypes=',
		parentKey: 'phenotypes',
		showTagType: false,
	}}
/>

<Canvas>
	<Story name='Tag'>
		{props => {
			const updateOnFilterBadge = () => {
				alert('removed');
			};
			return <Tag {...props} updateOnFilterBadgeHandler={updateOnFilterBadge} />;
		}}
	</Story>
</Canvas>
