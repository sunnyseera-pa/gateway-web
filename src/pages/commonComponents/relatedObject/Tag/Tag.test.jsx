import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { toTitleCase } from '../../../../utils/GeneralHelper.util';
import Tag from './Tag';

const props = {
	updateOnFilterBadgeHandler: jest.fn(),
	tagName: 'test',
	tagType: 'phenotype',
	onSearchPage: false,
	activeLink: false,
	url: '/search?search=&tab=Datasets&phenotypes=',
	parentKey: 'phenotypes',
	showTagType: false,
	filter: 'phenotypes',
};
let wrapper;

describe('Given the Tag component', () => {
	describe('When it is rendered', () => {
		beforeAll(() => {
			wrapper = render(<Tag {...props} />);
		});

		it('Then matches the previous snapshot', () => {
			expect(wrapper.container).toMatchSnapshot();
		});

		it(`Then Button should be rendered with className badge-${props.tagType}`, () => {
			expect(screen.getByTestId(`badge-${props.tagName}`)).toBeTruthy();
			expect(screen.getByTestId(`badge-${props.tagName}`)).toHaveClass(`badge-${props.tagType}`);
		});
	});
	describe('And activeLink is true', () => {
		it('Then the Tag without tagID should be rendered with links', () => {
			const { rerender } = wrapper;
			rerender(<Tag {...props} activeLink={true} />);
			expect(screen.getByTestId(`badge-${props.tagName}`)).toBeTruthy();
			expect(screen.getByTestId(`badge-${props.tagName}-link`)).toHaveAttribute('href', `${props.url}${props.tagName}`);
		});
		it('Then with tagID Tag should be rendered with id links', () => {
			const { rerender } = wrapper;
			rerender(<Tag {...props} activeLink={true} tagId='123' />);
			expect(screen.getByTestId(`badge-${props.tagName}`)).toBeTruthy();
			expect(screen.getByTestId(`badge-${props.tagName}-link`)).toHaveAttribute('href', `${props.url}1234`);
		});
	});

	describe('And showTagType is true', () => {
		it('Then the Tag  should be rendered with the prefix as TagType ', () => {
			const { rerender } = wrapper;
			rerender(<Tag {...props} activeLink={true} showTagType={true} />);
			expect(screen.getByTestId(`badge-${props.tagName}`)).toBeTruthy();
			expect(screen.getByTestId(`badge-${props.tagName}-link`)).toHaveTextContent(`${toTitleCase(props.tagType)}: ${props.tagName}`);
		});
	});

	describe('And activeLink and onSearchPage is true', () => {
		let updateOnFilterBadgeHandler = props.updateOnFilterBadgeHandler;
		it('Then Tag  should be rendered without links', () => {
			const { rerender } = wrapper;
			rerender(<Tag {...props} activeLink={true} onSearchPage={true} />);
			expect(screen.getByTestId(`badge-${props.tagName}`)).toBeTruthy();
			expect(screen.queryByTestId(`badge-${props.tagName}-link`)).toBeNull();
		});
		it('Then onclick Tag updateOnFilterBadgeHandler should be called', () => {
			fireEvent.click(screen.getByTestId(`badge-${props.tagName}`));
			expect(updateOnFilterBadgeHandler.mock.calls.length).toBe(1);
			expect(updateOnFilterBadgeHandler.mock.calls[0][0]).toEqual(props.parentKey);
			expect(updateOnFilterBadgeHandler.mock.calls[0][1]).toEqual({ label: props.tagName, parentKey: props.parentKey });
		});
	});
});
