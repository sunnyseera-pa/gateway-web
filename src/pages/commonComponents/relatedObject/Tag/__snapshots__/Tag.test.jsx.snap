// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Tag component When it is rendered Then matches the previous snapshot 1`] = `
<div>
  <div
    class="badge-phenotype"
    data-testid="badge-test"
  >
    test
  </div>
</div>
`;
